src/components/encrypted-intent.ts

```diff
@@ -0,0 +1,237 @@
+import { UserMembership } from "./membership-cache";
+import { APPSERVICE_LOGIN_TYPE, ClientEncryptionSession } from "./encryption";
+import Logging from "./logging";
+import { ReadStream } from "fs";
+import BotSdk, { MatrixClient } from "matrix-bot-sdk";
+import { FileUploadOpts, Intent, IntentOpts } from "./intent";
+
+const log = Logging.get("EncryptedIntent");
+
+export interface EncryptedIntentOpts {
+    sessionPromise: Promise<ClientEncryptionSession|null>;
+    sessionCreatedCallback: (session: ClientEncryptionSession) => Promise<void>;
+    ensureClientSyncingCallback: () => Promise<void>;
+    originalHomeserverUrl: string;
+}
+
+/**
+ * Implements some special handling on top of Intent to handle encrypted rooms.
+ */
+export class EncryptedIntent extends Intent {
+
+    private readonly encryptedRooms = new Map<string, string|false>();
+    private encryptionReadyPromise?: Promise<void>;
+    // A client that talks directly to the homeserver, bypassing pantalaimon.
+    private encryptionHsClient: MatrixClient;
+
+    constructor(
+        botSdkIntent: BotSdk.Intent,
+        botClient: BotSdk.MatrixClient,
+        intentOpts: IntentOpts, private encryptionOpts: EncryptedIntentOpts) {
+        super(botSdkIntent, botClient, intentOpts);
+
+        // We still need a direct client to the homeserver in some cases, so clone
+        // the existing one.
+        this.encryptionHsClient = new MatrixClient(
+            this.encryptionOpts.originalHomeserverUrl,
+            this.botSdkIntent.underlyingClient.accessToken
+        );
+        this.encryptionHsClient.impersonateUserId(this.userId);
+    }
+
+    /**
+     * Upload a file to the homeserver.
+     * @param content The file contents
+     * @param opts Additional options for the upload.
+     * @returns A MXC URL pointing to the uploaded data.
+     */
+    public async uploadContent(content: Buffer|string|ReadStream, opts: FileUploadOpts = {}): Promise<string> {
+        await this.ensureRegistered();
+        // Media is encrypted, since we don't know the destination room assume this media will be encrypted.
+        await this.encryptionOpts.ensureClientSyncingCallback();
+        return super.uploadContent(content, opts);
+    }
+
+    public onEvent(event: {
+        type: string,
```

Comment:
I'm suprised that RoomEvent from bot-sdk doesn't cover this (no room_id in it seemingly). I wonder if we ought to just define our type for this, since it comes up quite a bit ime.